src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -249,73 +271,66 @@ private void UpdateEndpoints()
-                    var part = segment.Parts[j];
-
-                    if (part.IsParameter &&
-                        part is RoutePatternParameterPart parameterPart &&
-                        action.RouteValues.ContainsKey(parameterPart.Name))
-                    {
-                        if (segmentParts == null)
-                        {
-                            segmentParts = segment.Parts.ToList();
-                        }
-                        if (allParameterPolicies == null)
-                        {
-                            allParameterPolicies = MvcEndpointInfo.BuildParameterPolicies(routePattern.Parameters, _parameterPolicyFactory);
-                        }
-
-                        // Replace parameter with literal value
-                        var parameterRouteValue = action.RouteValues[parameterPart.Name];
+                UpdatePathSegmentsWithLiterals(i, action, routePattern, newPathSegments, ref allParameterPolicies);
+            }
 
-                        // Route value could be null if it is a "known" route value.
-                        // Do not use the null value to de-normalize the route pattern,
-                        // instead leave the parameter unchanged.
-                        // e.g.
-                        //     RouteValues will contain a null "page" value if there are Razor pages
-                        //     Skip replacing the {page} parameter
-                        if (parameterRouteValue != null)
-                        {
-                            if (allParameterPolicies.TryGetValue(parameterPart.Name, out var parameterPolicies))
-                            {
-                                // Check if the parameter has a transformer policy
-                                // Use the first transformer policy
-                                for (var k = 0; k < parameterPolicies.Count; k++)
-                                {
-                                    if (parameterPolicies[k] is IOutboundParameterTransformer parameterTransformer)
-                                    {
-                                        parameterRouteValue = parameterTransformer.TransformOutbound(parameterRouteValue);
-                                        break;
-                                    }
-                                }
-                            }
+            IEnumerable<RoutePatternPathSegment> finalPathSegments = newPathSegments;
 
-                            segmentParts[j] = RoutePatternFactory.LiteralPart(parameterRouteValue);
-                        }
-                    }
+            if (requiredParameterEndpointRouteOrder != null)
+            {
+                // There are 3 cases where a parameter part can come into play:
+                // - /Home/{id?} - trailing single segment optional parameter
+                // - /Home.{id?} - trailing optional extension
+                // - /Home/{id?}/{catchAll*} - optional parameter then catch all
```

Comment:
Have you thought about what happens with multiple parameters?

Reply:
Can you have multiple optional parameters? I thought that wasn't allowed. I'll double check.

Reply:
You can have them. Will add support 👍